integration/route-ranking-test.ts

```diff
@@ -0,0 +1,162 @@
+          }
+
+          export default function Component() {
+            let data = useLoaderData();
+            return (
+              <>
+                <h1>:locale?/apple-pay</h1>
+                <pre>{data}</pre>
+              </>
+            );
+          }
+        `,
+        "app/routes/($locale).legal.$article.jsx": js`
+          import { useLoaderData } from '@remix-run/react';
+
+          export function loader() {
+            return ":locale?/legal/:article"
+          }
+
+          export default function Component() {
+            let data = useLoaderData();
+            return (
+              <>
+                <h1>:locale?/legal/:article</h1>
+                <pre>{data}</pre>
+              </>
+            );
+          }
+        `,
+      },
+      page
+    );
+
+    await app.goto("/legal/apple-pay");
+    await page.waitForSelector("h1");
+    expect(await app.getHtml("h1")).toMatch(":locale?/apple-pay");
+    expect(await app.getHtml("pre")).toMatch(":locale?/apple-pay");
+  });
+
+  test("preserves user-ordering from remix.config.js", async ({ page }) => {
+    let { app } = await setup(
+      {
+        "app/root.tsx": ROOT_FILE_CONTENTS,
+        "remix.config.js": js`
+          export default {
+            ignoredRouteFiles: [
+              "**/.*",
+              "routes/($locale).apple-pay.jsx",
+              "routes/($locale).legal.$article.jsx",
+            ],
+            async routes(defineRoutes) {
+              return defineRoutes((route) => {
+                route("/:locale?/legal/:article", "routes/($locale).legal.$article.jsx");
+                route("/:locale?/apple-pay", "routes/($locale).apple-pay.jsx");
+              });
+            },
+          };
+        `,
+        "app/routes/($locale).apple-pay.jsx": js`
+          import { useLoaderData } from '@remix-run/react';
+
+          export function loader() {
+            return ":locale?/apple-pay"
+          }
+
+          export default function Component() {
+            let data = useLoaderData();
+            return (
+              <>
+                <h1>:locale?/apple-pay</h1>
+                <pre>{data}</pre>
+              </>
+            );
+          }
+        `,
+        "app/routes/($locale).legal.$article.jsx": js`
+          import { useLoaderData } from '@remix-run/react';
+
+          export function loader() {
+            return ":locale?/legal/:article"
+          }
+
+          export default function Component() {
+            let data = useLoaderData();
+            return (
+              <>
+                <h1>:locale?/legal/:article</h1>
+                <pre>{data}</pre>
+              </>
+            );
+          }
+        `,
+      },
+      page
+    );
+
+    await app.goto("/legal/apple-pay");
+    await page.waitForSelector("h1");
+    expect(await app.getHtml("h1")).toMatch(":locale?/legal/:article");
+    expect(await app.getHtml("pre")).toMatch(":locale?/legal/:article");
```

Comment:
In this test, we expect `($locale).legal.$article.jsx` to win because we included it first in `remix.config.js` routes